pkg/oc/bootstrap/clusterup/components/persistent-volumes/setup_persistent_volumes.go

```diff
@@ -0,0 +1,249 @@
+	kbatch "k8s.io/kubernetes/pkg/apis/batch"
+	kapi "k8s.io/kubernetes/pkg/apis/core"
+	kubernetes "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait"
+
+	authorizationtypedclient "github.com/openshift/origin/pkg/authorization/generated/internalclientset/typed/authorization/internalversion"
+	"github.com/openshift/origin/pkg/oc/bootstrap/clusterup/componentinstall"
+	"github.com/openshift/origin/pkg/oc/bootstrap/docker/dockerhelper"
+	securityclient "github.com/openshift/origin/pkg/security/generated/internalclientset"
+	securitytypedclient "github.com/openshift/origin/pkg/security/generated/internalclientset/typed/security/internalversion"
+)
+
+const (
+	pvCount            = 100
+	pvSetupJobName     = "persistent-volume-setup"
+	pvInstallerSA      = "pvinstaller"
+	pvTargetNamespace  = "default"
+	pvIgnoreMarkerFile = ".skip_pv"
+)
+
+// TODO: This really should be a Go template
+const createPVScript = `#/bin/bash -e
+function generate_pv() {
+  local basedir="${1}"
+  local name="${2}"
+cat <<EOF
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: ${name}
+  labels:
+    volume: ${name}
+spec:
+  capacity:
+    storage: 100Gi
+  accessModes:
+    - ReadWriteOnce
+    - ReadWriteMany
+    - ReadOnlyMany
+  hostPath:
+    path: ${basedir}/${name}
+  persistentVolumeReclaimPolicy: Recycle
+EOF
+}
+
+function setup_pv_dir() {
+  local dir="${1}"
+  if [[ ! -d "${dir}" ]]; then
+    mkdir -p "${dir}"
+  fi
+  if ! chcon -t svirt_sandbox_file_t "${dir}" &> /dev/null; then
+    echo "Not setting SELinux content for ${dir}"
+  fi
+  chmod 770 "${dir}"
+}
+
+function create_pv() {
+  local basedir="${1}"
+  local name="${2}"
+
+  setup_pv_dir "${basedir}/${name}"
+  if ! oc get pv "${name}" &> /dev/null; then 
+    generate_pv "${basedir}" "${name}" | oc create -f -
+  else
+    echo "persistentvolume ${name} already exists"
+  fi
+}
+
+basedir="%[2]s"
+setup_pv_dir "${basedir}/registry"
+
+for i in $(seq -f "%%04g" 1 %[1]d); do
+  create_pv "${basedir}" "pv${i}"
+done
+`
+
+type SetupPersistentVolumesOptions struct {
+	InstallContext componentinstall.Context
+}
+
+func (c *SetupPersistentVolumesOptions) Name() string {
+	return "persistent-volumes"
+}
+
+func (c *SetupPersistentVolumesOptions) Install(dockerClient dockerhelper.Interface, logdir string) error {
+	kclient, err := kubernetes.NewForConfig(c.InstallContext.ClusterAdminClientConfig())
+	if err != nil {
+		return err
+	}
+	securityClient, err := securityclient.NewForConfig(c.InstallContext.ClusterAdminClientConfig())
+	if err != nil {
+		return err
+	}
+	authorizationClient, err := authorizationtypedclient.NewForConfig(c.InstallContext.ClusterAdminClientConfig())
+	if err != nil {
+		return err
+	}
+	if err := ensurePVInstallerSA(authorizationClient, kclient, securityClient); err != nil {
+		return err
+	}
+
+	_, err = kclient.Batch().Jobs(pvTargetNamespace).Get(pvSetupJobName, metav1.GetOptions{})
```

Comment:
Bad signal.  Make a TODO to make the job itself idempotent and non-failing.

Reply:
I'll have a closer look into it once it merges. We need to ensure that strong idempotency.